File: ui/src/test/java/com/qacart/todo/pages/SignupPage.java
package com.qacart.todo.pages;

import com.qacart.todo.base.BasePage;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class SignupPage extends BasePage {

    public SignupPage(WebDriver driver) {
        super(driver);
    }

    private final By signupBtn = By.cssSelector("[data-testid=signup]");
    private final By firstName = By.cssSelector("[data-testid=first-name]");
    private final By lastName = By.cssSelector("[data-testid=last-name]");
    private final By email = By.cssSelector("[data-testid=email]");
    private final By password = By.cssSelector("[data-testid=password]");
    private final By confirmPassword = By.cssSelector("[data-testid=confirm-password]");
    private final By submit = By.cssSelector("[data-testid=submit]");
    private final By goLogin = By.cssSelector("[data-testid=go-login]");

    public void clickSignup() {
        driver.findElement(signupBtn).click();
    }

    public void fillSignupForm(String fn, String ln, String em, String pw, String cpw) {
        driver.findElement(firstName).clear();
        driver.findElement(firstName).sendKeys(fn);

        driver.findElement(lastName).clear();
        driver.findElement(lastName).sendKeys(ln);

        driver.findElement(email).clear();
        driver.findElement(email).sendKeys(em);

        driver.findElement(password).clear();
        driver.findElement(password).sendKeys(pw);

        driver.findElement(confirmPassword).clear();
        driver.findElement(confirmPassword).sendKeys(cpw);
    }

    public void submitSignup() {
        driver.findElement(submit).click();
    }

    public void signup(String fn, String ln, String em, String pw, String cpw) {
        clickSignup();
        fillSignupForm(fn, ln, em, pw, cpw);
        submitSignup();
    }

    public boolean isMessageDisplayed() {
        return driver.findElement(goLogin).isDisplayed();
    }

    public boolean isSignupFormDisplayed() {
        return driver.findElement(firstName).isDisplayed();
    }
}

File: ui_cypress_bdd/cypress/support/pages/SignupPage.js
const { BasePage } = require('./BasePage');

/**
 * Équivalent Cypress de src/pages/Id01_SignupPage.ts (ui_playwright_bdd).
 * Le fallback multi-sélecteur Playwright (getByTestId().or(css).or(placeholder))
 * devient un sélecteur CSS unique séparé par virgules (sémantique OR).
 */
class SignupPage extends BasePage {
  open() {
    this.goto('/signup');
  }

  fillSignupForm(firstName, lastName, email, password, confirmPassword) {
    const confirm = confirmPassword ?? password;
    this._fillField('[data-testid="first-name"], input[name="firstName"], [placeholder*="first name" i]', firstName);
    this._fillField('[data-testid="last-name"], input[name="lastName"], [placeholder*="last name" i]', lastName);
    this._fillField('[data-testid="email"], input[name="email"], [placeholder*="email" i]', email);
    this._fillField('[data-testid="password"], input[name="password"], [placeholder="password" i]', password);
    this._fillField('[data-testid="confirm-password"], input[name="confirmPassword"], [placeholder*="confirm password" i]', confirm);
  }

  // cy.type() refuse les chaînes vides — on se contente du clear() pour simuler un champ laissé vide
  _fillField(selector, value) {
    const field = cy.get(selector).first().should('be.visible').clear();
    if (value) cy.get(selector).first().type(value);
    return field;
  }

  signup(firstName, lastName, email, password, confirmPassword) {
    this.fillSignupForm(firstName, lastName, email, password, confirmPassword);
    cy.get('[data-testid="submit"], button').contains(/sign up|signup|register|create account/i).click();
  }

  assertSignupError(expectedText) {
    const pattern = expectedText || /invalid|required|missing|weak|mismatch|already registered|email/i;
    cy.contains(pattern, { timeout: 15000 }).should('be.visible');
  }

  assertSignedUp() {
    cy.url({ timeout: 10000 }).should('match', /dashboard|home|todo|tasks/i);
  }
}

module.exports = { SignupPage };
